### ISSUE TITLE
Rating not available when path to file differs from library path

### ISSUE THREAD

**Issue by mohicanb** [clementine-player/Clementine#5351]
### Before posting
issue #5065 may be a duplicate of this one.

### System information

- Operating System: Ubuntu 14.04
- Clementine version: 1.2 (from Ubuntu repositories) and 1.3 (from ppa)

### Expected behaviour / actual behaviour
For some files, under some circumstances, **when adding them from the disk files**, rating is not available. I also observed that in this case automatic tags such as last played date, number of listening and score, do not update.

Sometimes this only occurs when adding with drag and drop, sometimes it occurs also when adding from the context menu.

### Steps to reproduce the problem (only for bugs)

Summary: the problem occurs when the path to the file, displayed on the playlist, column 'file (with path)' is not included in the library folders.
This means that **it occurs both for files that are not included in the library and for files that are accessed via another path (through symbolic links).**
Clementine behaviour is different when adding a file with the context menu: it adds the file with direct path to disk file, and when adding with drag and drop: it adds the file with the path specified in the file browser box (this may use symbolic links).

1/ store your music files in different disk partition that /
e.g. /media/DATA/username/Music/musicfile.ogg
(or put it on a USB stick)

2/ erase (or rename) the /home/username/Music folder and add a symbolic link in your /home/username/ folder that points to the /media/DATA/username/Music folder, give it the name 'Music'

3/ configure your Clementine library: add folder /home/username/Music
The resulting path will be added: /media/DATA/username/Music

4/ BROWSE USING SYMBOLIC LINK
in the file tab, browse to the path /home/username/Music/

4.1/ add the musicfile.ogg to a playlist with the context menu
-> the path in the playlist will be: '/media/DATA/...', the rating will be available

4.2/ add the musicfile.ogg to a playlist with drag and drop
-> the path in the playlist will be: '/home/...', the rating will NOT be available
(I guess because Clementine will not recognise that the file belongs to the library)
**This is a bug.**

5/ BROWSE DIRECT PATH 
in the file tab, browse to the path /media/DATA/username/Music/

5.1/ add the musicfile.ogg to a playlist with the context menu
-> the path in the playlist will be: '/media/DATA/...', the rating will be available

5.2/ add the musicfile.ogg to a playlist with the context menu
-> the path in the playlist will be: '/media/DATA/...', the rating will be available

6/ FILE NOT IN LIBRARY (this may not be considered a bug)

copy your musicfile.ogg in the /media/DATA/username/Documents folder

in the file tab, browse to the direct path /media/DATA/username/Documents/

6.1/ add the musicfile.ogg to a playlist with the context menu
-> the path in the playlist will be: '/media/DATA/...', the rating will NOT be available

6.2/ add the musicfile.ogg to a playlist with drag and drop
-> the path in the playlist will be: '/media/DATA/...', the rating will NOT be available

**Comment by mohicanb** [comment]
Still another way to reproduce the bug, even when adding a track with the context menu.

7.1/ store a music file in another partition
e.g. /media/DATA2/MoreMusic/

7.2/ add a symbolic link in the /home/username/Music folder to the /media/DATA2/MoreMusic/, call it 'MoreMusic'

Therefore the files are included in the library you configured in point 3 (see previous post). Because 
the library folder is /media/DATA/username/Music, wich includes /media/DATA/username/Music/MoreMusic

7.3/ BROWSE USING SYMBOLIC LINK
in the file tab, browse to the path /home/username/Music/MoreMusic/

7.3.1/ add the musicfile.ogg to a playlist with the context menu
-> the path in the playlist will be: '/media/DATA2/...', the rating will NOT be available
**This is a bug.**

7.3.2/ add the musicfile.ogg to a playlist with drag and drop
-> the path in the playlist will be: '/home/...', the rating will NOT be available
**This is a bug.**

**Comment by hatstand** [comment]
Ratings are only intended to work for the library.

**Comment by mohicanb** [comment]
Sorry you didn't read completely: the bug also occurs for files which ARE included in the library. In the case these files are accessed through symbolic links.

Please reopen !

**Comment by Ferroin** [comment]
The point is you're trying to access things through paths outside the library and expecting all the features to work, which is not ever going to be the case.  Clementine doesn't parse symlinks, it lets the filesystem do that (which is exactly what it should do, because special handling of symlinks causes all kinds of issues), so it doesn't know that the file you are pointing it at is a symlink to something that's in your music library, it just sees a audio file that it isn't managing.

You should be pointing Clementine at the path you actually intend to access the files from.  For example, I keep my music on Dropbox, and have my Dropbox folder in a separate filesystem (so I can share the disk space between Windows and Linux on my laptop, and so I can use different storage parameters on my desktop).  The 'Music' folder in my home directory on both systems is a symbolic link pointing to the appropriate folder in my Dropbox folder.  I point Clementine and everything else at my Music folder via the symlink, not straight at the folder in Dropbox, because that's how I access my music (and how I expect it to be reported in saved playlists).

**Comment by hatstand** [comment]
tl;dr: Just use the library

**Comment by mohicanb** [comment]
@hatsland:

browsing the files from the directory structure is sometimes useful ! This is why Clementine is better than Rhythmbox - for example - which does not have this ability.

@Ferroin:

1. yet the behaviour of Clementine related to symlinks is different when adding a file with the context menu or with drag and drop. Why ? **This inconsistency should be considered a bug.**

2. your example with Dropbox is precisely what I am doing too: have the files on different filesystems and use symlinks from the /home folder. So you should experience the same bug when browsing from your /home folder and adding your Dropbox files to a playlist. At least with the drag and drop method.

---
My best guess for a solution would be that Clementine preserve paths using symbolic links all the time,
- when configuring the library
- when adding the files to the playlist using the context menu
- when loading playlists including path that use symbolic links
- etc.

If the path in the library configuration would be set to /home/username/Music (instead of the direct disk path /media/DATA/username/Music in my previous example) then, since browsing is easier to do starting the /home/username folder (thanks to the 'home' button) then all files that will be drag and dropped to the playlist would be detected as part of the library.

If the context menu also preserves the path using symbolic links then it would also work with this adding method, and the path in the playlist would be consistent.

This consistency is also important when saving playlists to the disk. And when loading them.

**Comment by Ferroin** [comment]
We do appear to be parsing symlinks when adding directories to the library.  I somehow had not noticed this before.  I'm curious to hear from one of the developers what the specific reasoning behind this behavior is, as it's debatably incorrect behavior, and can lead to confusion (the only end-user applications that should be parsing symlinks are things that need to store the exact state of the filesystem (like backup tools or VCS software)).

At the very least, we need to change how we're handling things so it's consistent, either we should always parse symlinks ourselves, or we never should.

**Comment by Ferroin** [comment]
Also, to further clarify, this is broken behavior, but it's because of an inconsistency in how Clementine is handling symbolic links in paths.

Based on some minimal testing:
    * Adding directories to the Library canonicalizes the path removing any symbolic links
    * Adding files from the file manager pane, or from drag'n'drop does not canonicalize the path, but still follows symlinks.

So, in my setup, when I add the folder /home/austin/Music to my Library, it ends up pointing at /mnt/dropbox/Dropbox/Music instead, but when I try to add a file from the file pane with a path of '/home/austin/Music/some_file.flac', it gets added to the playlist with that name, so Clementine doesn't think it's in the Library, even though it is.

**Comment by chrisgraham** [comment]
> The point is you're trying to access things through paths outside the library and expecting all the features to work, which is not ever going to be the case. Clementine doesn't parse symlinks, it lets the filesystem do that (which is exactly what it should do, because special handling of symlinks causes all kinds of issues), so it doesn't know that the file you are pointing it at is a symlink to something that's in your music library, it just sees a audio file that it isn't managing.
> 
> You should be pointing Clementine at the path you actually intend to access the files from. For example, I keep my music on Dropbox, and have my Dropbox folder in a separate filesystem (so I can share the disk space between Windows and Linux on my laptop, and so I can use different storage parameters on my desktop). The 'Music' folder in my home directory on both systems is a symbolic link pointing to the appropriate folder in my Dropbox folder. I point Clementine and everything else at my Music folder via the symlink, not straight at the folder in Dropbox, because that's how I access my music (and how I expect it to be reported in saved playlists).

The problem here is sometimes people are having to be doing special things due to low disk space. In my case I have an SSD that is 250GB, but it's actually 2 SSDs in one case so I have 2 partitions. Holding my music library on just one isn't feasible, so I split it up using some symlinks for individual artist directories.
I know I could try and do something smarter like fuse the drives together in my mount configuration, but I think what I'm doing should in theory be more stable and reliable.
